### ISSUE TITLE
jQuery Typeahead Search from json file

### ISSUE THREAD

**Issue by TheGreenArrow** [running-coder/jquery-typeahead#234]
I'm trying to create an autocomplete search based on json file.
I need the **beer v1** example in the [demo page](http://www.runningcoder.org/jquerytypeahead/demo/).

I have the html, js code and a json file with this format : 
["CANILLO","ENCAMP","ESCALDES-ENGORDANY"]

I can't link the file to the source to make it work .
I think the problem is here  : 

>     source: {
        "ale": {
            ajax: {
                url: "/jquerytypeahead/beer_v1.json",
                path: "data.beer.ale"
            }
        },

**Comment by running-coder** [comment]
would you be able to put something up with jsfiddle? Its hard for me to understand your setup and whats missing.. or you can paste me everything and we will go from there

**Comment by TheGreenArrow** [comment]
This is the code i have for now. I need to add a json file for every group of results.

    `<link rel="stylesheet" href="../src/jquery.typeahead.css">
    <script src="http://code.jquery.com/jquery-2.1.0.min.js"></script>
    
    <script src="../src/jquery.typeahead.js"></script>
    <div style="width: 100%; max-width: 500px; margin: 0 auto;">
    <h1>Demo</h1>
    <div class="js-result-container"></div>
    <form>
        <div class="typeahead__container">
            <div class="typeahead__field">
            <span class="typeahead__query">
                <input class="js-typeahead-input"
                       name="q"
                       type="search"
                       autofocus
                       autocomplete="off">
            </span>
            </div>
        </div>
    </form>

    <script>
        var data = {
            countries: ["Afghanistan", "Albania", "Algeria"],
            capitals: ["Abu Dhabi", "Abuja", "Accra"]
        };
        typeof $.typeahead === 'function' && $.typeahead({
            input: ".js-typeahead-input",
            minLength: 1,
            order: "asc",
            group: true,
            maxItemPerGroup: 5,
            groupOrder: function () {
                var scope = this,
                    sortGroup = [];

                for (var i in this.result) {
                    sortGroup.push({
                        group: i,
                        length: this.result[i].length
                    });
                }
                sortGroup.sort(
                    scope.helper.sort(
                        ["length"],
                        true, // false = desc, the most results on top
                        function (a) {
                            return a.toString().toUpperCase()
                        }
                    )
                );
                return $.map(sortGroup, function (val, i) {
                    return val.group
                });
            },
            hint: true,
            dropdownFilter: "Tous",
            template: "{{display}}, <small><em>{{group}}</em></small>",
            emptyTemplate: "pas de résultats pour : {{query}}",
            source: {
                pays: {
                    data: data.countries
                },
                capitale: {
                    data: data.capitals
                }
            },
            debug: true
        });
    </script>
    </div>`

**Comment by TheGreenArrow** [comment]
Any help here please ?

**Comment by running-coder** [comment]
hey sorry for the delay, so you can either reference your file like this:

```
source: {
    pays: {
        ajax: {
            url: './hello.json'
        }
    }
}
```
or use a more ES6 way to do it using webpack or another tool and import your json
```
import myJson from 'json!./hello.json';

...

source: {
    pays: {
        data: myJson
    }
}
```

**Comment by TheGreenArrow** [comment]
Well, i tried this code but it doesn't work : 

`<link rel="stylesheet" href="../src/jquery.typeahead.css">
<script src="http://code.jquery.com/jquery-2.1.0.min.js"></script>
<script src="../src/jquery.typeahead.js"></script>

<div style="width: 500px; margin: 0 auto;">
    <h1>Country_v2 Demo</h1>
	    <div class="js-result-container"></div>
    <form>
        <div class="typeahead__container">
            <div class="typeahead__field">
            <span class="typeahead__query">
                <input class="js-typeahead-input"
                       name="q"
                       type="search"
                       autofocus
                       autocomplete="off">
            </span>
            <span class="typeahead__button">
                <button type="submit">
                    <span class="typeahead__search-icon"></span>
                </button>
            </span>
            </div>
        </div>
    </form>

    <script>
        typeof $.typeahead === 'function' && $.typeahead({
            input: ".js-typeahead-input",
            minLength: 1,
            order: "asc",
            group: true,
            maxItemPerGroup: 3,
            groupOrder: function () {
                var scope = this,
                    sortGroup = [];
                for (var i in this.result) {
                    sortGroup.push({
                        group: i,
                        length: this.result[i].length
                    });
                }
                sortGroup.sort(
                    scope.helper.sort(
                        ["length"],
                        false, // false = desc, the most results on top
                        function (a) {
                            return a.toString().toUpperCase()
                        }
                    )
                );
                return $.map(sortGroup, function (val, i) {
                    return val.group
                });
            },
            hint: true,
            dropdownFilter: "All",
            template: "{{display}}, <small><em>{{group}}</em></small>",
            emptyTemplate: "no result for {{query}}",
			source: {
				pays: {
					ajax: {
						url: 'villes.json'
					}
				}
			},
            debug: true
        });
    </script>
</div>`

**Sample of villes.json**

`["CANILLO","ENCAMP","ESCALDES-ENGORDANY","LA FARGA DE MOLES","AIXOVALL"]`

**Comment by running-coder** [comment]
works for me, verify that villes.json doesn't produce a 404 in your network tab, you may need to re-adjust it's location inside `ajax.url`

![screen shot 2016-09-16 at 3 09 50 pm](https://cloud.githubusercontent.com/assets/7377382/18598215/af11e864-7c1f-11e6-91d4-e33bc6936f95.png)

**Comment by TheGreenArrow** [comment]
Thank you, it's working now. The problme was in the file villes.json which contains a null values that makes the search load forever. Can we fix that and skip null if exists?
Another question please : 
I changed the json file to add more data : 

`[{"Libelle":"CANILLO","ID":2,"Koedia":"ADCAN","Pays":"ANDORRE"},{"Libelle":"ENCAMP","ID":3,"Koedia":"ADENC","Pays":"ANDORRE"},{"Libelle":"ESCALDES-ENGORDANY","ID":4,"Koedia":"ADESC","Pays":"ANDORRE"},{"Libelle":"LA FARGA DE MOLES","ID":5,"Koedia":"ADFMO","Pays":"ANDORRE"}]`

I need to put **ID** and **Koedia** values in 2 seperated hidden inputs to use them in the form, **Pays** value should be added in every result item to remplace group name. For example : 

![screenshot 2016-09-19](https://cloud.githubusercontent.com/assets/21695752/18626576/d3c0f602-7e4d-11e6-9850-7052e6b1d268.png)

**Comment by TheGreenArrow** [comment]
Any help here please ?

**Comment by running-coder** [comment]
hey, to place the values in other hidden input you can use the onclick callback
```
onclick: function (node, a, item, event) {
    $('#hiddenInput').val(item.ID)
}
```
to alter the display you can use your own `template` and specify which field you want. I recommend looking at the options.template doc and examples.

**Comment by TheGreenArrow** [comment]
I made some changes : 

```
<link rel="stylesheet" href="jquery.typeahead.css">
<script src="http://code.jquery.com/jquery-2.1.0.min.js"></script>
<script src="jquery.typeahead.js"></script>
<div style="max-width: 500px; margin: 0 auto;">
    <h1>Demo</h1>
    <form>
        <div class="typeahead__container">
            <div class="typeahead__field">
            <span class="typeahead__query">
                <input class="js-typeahead-input" name="q" type="search" autofocus autocomplete="off">
            </span>
            </div>
        </div>
                <input id="ID" type="text" >
                <input id="KOEDIA" type="text" >
</form>
    <script>
        typeof $.typeahead === 'function' && $.typeahead({
            input: ".js-typeahead-input",
            minLength: 1,
			maxItem: 40,
			order: "asc",
            group: true,
            maxItemPerGroup: 10,
            groupOrder: function () {
                var scope = this,
                    sortGroup = [];
                for (var i in this.result) {
                    sortGroup.push({
                        group: i,
                        length: this.result[i].length
                    });
                }
                sortGroup.sort(
                    scope.helper.sort(
                        ["length"],
                        false, // false = desc, the most results on top
                        function (a) {
                            return a.toString().toUpperCase()
                        }
                    )
                );
                return $.map(sortGroup, function (val, i) {
                    return val.group
                });
            },
            hint: true,
            dropdownFilter: "Tous",
            // template: "{{display}}, <small><em>{{group}}</em></small>",
            template: function (query, item) {
                return '<span class="row">' +
                    '<span>{{item.Libelle}}</span>, <small><em>{{item.Pays}}</em></small>' +
                "</span>"
            },
            emptyTemplate: "aucun résultat pour : {{query}}",
            source: {
                regions: {
                    ajax: {
                        url: 'regions.json',
                        path: 'item.Libelle',
                    }
                },
                zones: {
                    ajax: {
                        url: 'zones.json',
                        path: 'item.Libelle',
                    }
                },
                villes: {
                    ajax: {
                        url: 'villes.json',
                        path: 'item.Libelle',
                    }
                },
                parcs: {
                    ajax: {
                        url: 'parcs.json',
                        path: 'item.Libelle',
                    }
                },
				},
            callback: {
                onClick: function (node, a, item, event) {
                    $('#ID').val(item.ID)
                    $('#KOEDIA').val(item.Koedia)
                },
            },
            debug: true
        });
    </script>
</div>
```
but it does'nt work... did i miss something ?

**Comment by running-coder** [comment]
can you paste me the output of `regions.json`

**Comment by TheGreenArrow** [comment]
All files have the same structure : 

`[{"Libelle":"CANILLO","ID":2,"Koedia":"ADCAN","Pays":"ANDORRE"},{"Libelle":"MADRID","ID":3,"Koedia":"ADENC","Pays":"SPAIN"},{"Libelle":"PARIS","ID":4,"Koedia":"ADESC","Pays":"FRANCE"},{"Libelle":"LONDON","ID":5,"Koedia":"ADFMO","Pays":"ENGLAND"}]`

**Comment by running-coder** [comment]
ok so remove the path from your ajax request, path is used in case your data is deep inside the response. You can also remove item. from your template

**Comment by TheGreenArrow** [comment]
Updated but still not working...

```
<link rel="stylesheet" href="jquery.typeahead.css">
<script src="http://code.jquery.com/jquery-2.1.0.min.js"></script>
<script src="jquery.typeahead.js"></script>

<div style="max-width: 500px; margin: 0 auto;">

    <h1>Demo</h1>

    <form>
        <div class="typeahead__container">
            <div class="typeahead__field">
            <span class="typeahead__query">
                <input class="js-typeahead-input" name="q" type="search" autofocus autocomplete="off">
            </span>
            </div>
        </div>
                <input id="ID" type="text" >
                <input id="KOEDIA" type="text" >
</form>

    <script>
        typeof $.typeahead === 'function' && $.typeahead({
            input: ".js-typeahead-input",
            minLength: 1,
			maxItem: 40,
			order: "asc",
            group: true,
            maxItemPerGroup: 10,
            groupOrder: function () {
                var scope = this,
                    sortGroup = [];
                for (var i in this.result) {
                    sortGroup.push({
                        group: i,
                        length: this.result[i].length
                    });
                }
                sortGroup.sort(
                    scope.helper.sort(
                        ["length"],
                        false, // false = desc, the most results on top
                        function (a) {
                            return a.toString().toUpperCase()
                        }
                    )
                );
                return $.map(sortGroup, function (val, i) {
                    return val.group
                });
            },
            hint: true,
            dropdownFilter: "Tous",
            template: function (query, item) {
                return '<span class="row">' +
                    '<span>{{Libelle}}</span>, <small><em>{{Pays}}</em></small>' +
                "</span>"
            },
            emptyTemplate: "aucun résultat pour : {{query}}",
            source: {
                regions: {
                    ajax: {
                        url: 'regions.json',
                        path: 'Libelle',
                    }
                },
                zones: {
                    ajax: {
                        url: 'zones.json',
                        path: 'Libelle',
                    }
                },
                villes: {
                    ajax: {
                        url: 'villes.json',
                        path: 'Libelle',
                    }
                },
                parcs: {
                    ajax: {
                        url: 'parcs.json',
                        path: 'Libelle',
                    }
                },
				},
            callback: {
                onClick: function (node, a, item, event) {
                    $('#ID').val(ID)
                    $('#KOEDIA').val(Koedia)
                },

            },
            debug: true

        });
    </script>
</div>
```

**Comment by running-coder** [comment]
remove all paths from ajax requests

**Comment by TheGreenArrow** [comment]
Updated and nothing

**Comment by running-coder** [comment]
you were missing the `display` and item.ID, `item.Koedia` inside the `onClick` function

https://jsfiddle.net/Lmnyhs4d/

**Comment by TheGreenArrow** [comment]
Almost perfect ! Just one last thing, how can exclude **Pays** from the search ?
![2016-09-26 15 02 10](https://cloud.githubusercontent.com/assets/21695752/18837041/a326f090-8402-11e6-9cb9-ce0ac7b11085.jpg)

**Comment by running-coder** [comment]
cool!, simply remove it from the `display`

**Comment by TheGreenArrow** [comment]
It's working fine with the included json data, but not with an external file.

**Comment by running-coder** [comment]
perhaps you can build an online demo and I can debug from there. At this point it all depends how you include / fetch the json files.

**Comment by TheGreenArrow** [comment]
Never mind, the problem was in the file which contains some null values (you script don't like null values).
Anyway, thank you so much !

**Comment by running-coder** [comment]
I've added a fix for that (following with your ticket) #239 , still on the develop branch will be in 2.7.1 (next release)

**Comment by running-coder** [comment]
you welcome!
